compat/src/main/scala-2.11_2.12/scala/collection/compat/CompatImpl.scala

```diff
@@ -14,36 +14,35 @@ package scala.collection.compat
 
 import scala.reflect.ClassTag
 import scala.collection.generic.CanBuildFrom
+import scala.{collection => c}
 import scala.collection.{immutable => i, mutable => m}
 
 /* builder optimized for a single ++= call, which returns identity on result if possible
  * and defers to the underlying builder if not.
  */
-private final class IdentityPreservingBuilder[A, CC[X] <: TraversableOnce[X]](
-    that: m.Builder[A, CC[A]])(implicit ct: ClassTag[CC[A]])
-    extends m.Builder[A, CC[A]] {
+private abstract class PreservingBuilder[A, C <: TraversableOnce[A]] extends m.Builder[A, C] {
+  val that: m.Builder[A, C]
+  val ct: ClassTag[C]
 
   //invariant: ruined => (collection == null)
-  var collection: CC[A] = null.asInstanceOf[CC[A]]
-  var ruined            = false
+  var collection: C = null.asInstanceOf[C]
+  var ruined        = false
 
   private[this] def ruin(): Unit = {
     if (collection != null) that ++= collection
-    collection = null.asInstanceOf[CC[A]]
+    collection = null.asInstanceOf[C]
     ruined = true
   }
 
   override def ++=(elems: TraversableOnce[A]): this.type =
     elems match {
```

Comment:
nit. I assume the `ct` unapply happens before the predicate check, but those are very fast and this branch can only be taken once. What about:
```scala
@inline def proxy(): Unit = {
  ruin()
  that ++= elems
}

if ((collection eq null) && !ruined) {
  elems match {
    case ct(ca) =>
      collection = ca 
    case _ => proxy()
}
else proxy()

this
```

Reply:
👍  thanks